### PR TITLE
Make KeyringProvider::fetch_* async

### LINKED ISSUE #3073: `KeyringProvider::fetch_subprocess` should be async
As far as I can tell, we're egregiously blocking the event loop for 80ms on keyring fetches. The function should be async and use an async subprocess API.

### PR DESCRIPTION
**Pull request by naosense**
To resolve #3073

### CODE REVIEW

File: crates/uv-auth/src/middleware.rs [hunk 1]
```diff
@@ -142,6 +142,7 @@ impl Middleware for AuthMiddleware {
                 .and_then(|credentials| credentials.username())
             {
                 debug!("Checking keyring for credentials for {url}");
+                // how to fix this
                 keyring.fetch(request.url(), username)
```
Comment (naosense) [inline crates/uv-auth/src/middleware.rs:146]: hi @zanieb  I googled for a while, but still no idea about how to edit this

Comment (naosense) [inline crates/uv-auth/src/middleware.rs:146]: hi @zanieb   I googled for a while, but still no idea about how to edit this

Comment (zanieb) [reply crates/uv-auth/src/middleware.rs:146]: I believe the idiomatic way to do it is with `match` instead

```diff
diff --git a/crates/uv-auth/src/middleware.rs b/crates/uv-auth/src/middleware.rs
index 191ca92e..451f690a 100644
--- a/crates/uv-auth/src/middleware.rs
+++ b/crates/uv-auth/src/middleware.rs
@@ -136,19 +136,25 @@ impl Middleware for AuthMiddleware {
         //      falls back to the host, but we cache the result per host so if a keyring
         //      implementation returns different credentials for different URLs in the
         //      same realm we will use the wrong credentials.
-        } else if let Some(credentials) = self.keyring.as_ref().and_then(|keyring| {
-            if let Some(username) = credentials
-                .get()
-                .and_then(|credentials| credentials.username())
-            {
-                debug!("Checking keyring for credentials for {url}");
-                // how to fix this
-                keyring.fetch(request.url(), username)
-            } else {
-                trace!("Skipping keyring lookup for {url} with no username");
-                None
+        } else if let Some(credentials) = match self.keyring {
+            Some(ref keyring) => {
+                match credentials
+                    .get()
+                    .and_then(|credentials| credentials.username())
+                {
+                    Some(username) => {
+                        debug!("Checking keyring for credentials for {url}");
+                        // how to fix this
+                        keyring.fetch(request.url(), username).await
+                    }
+                    None => {
+                        trace!("Skipping keyring lookup for {url} with no username");
+                        None
+                    }
+                }
             }
-        }) {
+            None => None,
+        } {
             debug!("Found credentials in keyring for {url}");
             request = credentials.authenticate(request);
             new_credentials = Some(Arc::new(credentials));
```

Then you'll run into problems with the `Mutex` being synchronous.

**Comment by zanieb** [comment]
Fyi we're chatting about using the tokio mutex over in https://github.com/astral-sh/uv/pull/3105 – it might not be necessary.

**Comment by naosense** [comment]
> Fyi we're chatting about using the tokio mutex over in #3105 – it might not be necessary.

Ok, i see. Now i split the cache lock to avoid across `await` call.

**Comment by naosense** [comment]
Hey @zanieb, i think this pr is ready to review, let me know if there's anything else i need to do.

File: crates/uv-auth/src/middleware.rs [hunk 2]
```diff
@@ -136,18 +136,25 @@ impl Middleware for AuthMiddleware {
...
-                None
+        } else if let Some(credentials) = match self.keyring {
+            Some(ref keyring) => {
+                match credentials
+                    .get()
+                    .and_then(|credentials| credentials.username())
+                {
+                    Some(username) => {
+                        debug!("Checking keyring for credentials for {url}");
+                        // how to fix this
```
[the remark this answers is not in the document]
Comment (naosense) [reply crates/uv-auth/src/middleware.rs]: sorry, forgot to remove

**Comment by zanieb** [comment]
Ah it looks like you have some conflicts with my changes in #3130, sorry about that lmk if you want me to fix it.

We also _might_ want to switch to the async mutex since we do want to hold across await points now, interesting. We're also considering switching to a `RwLock`. I think all that is probably worth doing/considering separately.

**Comment by naosense** [comment]
OK, thanks for you information, I think it makes sense to split these changes into multiple PR, what do you think?

**Comment by zanieb** [comment]
Thank you!

**Pull request merged**